docs/architecture/adr-043-interchain-identifiers.md

```diff
@@ -0,0 +1,108 @@
+
+IIDs also enable (off-chain) assertions to be made about (on-chain) digital assets – for instance, in the form of Verifiable Credentials.
+
+An IID Method DID is **self-referential**, which means the asset describes and locates itself in its chain namespace.
+
+Compare this with a generic, all-purpose DID which can be used to identify and refer to _any_ entity. For instance, the canonical application of DIDs is Self-Sovereign Identity, which provides people and other entities with identifiers together with cryptographic methods to authenticate as the subject and to exercise control over the state of the DID Document.
+
+---
+We define **Interchain Identifiers** as being explicitly purposed to identify and locate assets within a chain namespace.
+
+---
+
+- An asset on a Cosmos chain <-------- Cosmos IID identifies and locates
+- Cosmos DID --------> Identifies and locates an external thing
+
+All DIDs have an associated **DID Document**, which should subscribe to the W3C DID-Core standard.
+This defines the cryptographic control structure for a DID, and may also specify the Service types and endpoints which are related to the DID subject.
+
+The properties of a DID-Core standard DID Document include:
+- **Identifier** (the DID) of the subject with the Identifier of the Controller who is authorized to update the DID Document.
+- **Verificaton Method/s** (The cryptographic key material).
+- **Verification Relationship/s** (Describing and restricting the purposes for which the verification Methods can be used, such as Authentication).
+- **Services** related to the DID subject, with endpoints for reaching these services, such as data stores or blockchain services.
+
+For the IID Method, two additional DID Document properties have been specified, which are required for many types of on-chain assets:
+- **Linked Resources** for describing the resources which are associated with an on-chain asset, such as a media file together with its resource identifier, type, format, location and cryptographic proof (such as a hash). 
+- **Accorded Rights** for recording machine-executable capabilities, legal rights, or any other entitlements, which are accorded to the owner/controller of an asset.
+
+Linked Resources can demonstrate the distinctions between an IID and DID:
+```
+Asset IID   // self-referrant <--
+├── Linked Resources  // examples
+│   ├── Resource DID  // externally referrant -->
+│   ├── Other Resource URL
+│   ├── Another Resource CID (Content ID)
+```
+
+The **DID Syntax** for an IID conforms to the [DID-Core Syntax](), which would be expressed as follows for an asset minted from the NFT Module on a Cosmos chain:
+
+`DID:COSMOS:CHAIN:MODULE:abc123`
+
+Where: 
+- `COSMOS` is the method _(it is not the Cosmos Zone namespace)_.
+- `CHAIN` is the namespace, such as `IXO` or `REGEN` _(it is not the chain-id)_.
+- `MODULE` is the module namespace, such as `NFT` or `BANK` _(as named within the context of each application chain)_.
```

Comment:
So I think this is maybe where there are some misunderstandings arising. I have been advocating for namespaced denoms within bank. But `bank` would never itself issue any denoms itself but rather manage denoms from a number of sources that are likely namespaced. I would feel more comfortable using the language NAMESPACE instead of MODULE. Module is the software component and there likely won't be a 1:1 correspondence. Instead, a chain would have a mapping for which modules deal with which namespaces of assets IID. I think fundamentally this comes down to the SDK's information model for assets and identities - it's not as if just any module exposes its own asset interface. Currently only two modules fundamentally deal with assets and identities - bank and auth respectively. Maybe there will be a third one for NFTs if that doesn't get dealt with by bank, but either way it's a different level of the stack. The namespace, however, would likely point to a class of assets dealt with by a specific module or rule set proxied through the generic bank, nft and auth module APIs. Hope that makes sense.

Reply:
> I would feel more comfortable using the language NAMESPACE instead of MODULE.

@aaronc this makes complete sense and I will update accordingly.

>   It's not as if just any module exposes its own asset interface.

Agreed, but we should at least have a core set of standard interfaces for modules which deal with assets.
There's still a question in my mind about whether an asset module should store its own identifiers and metadata, vs relying on a an identifier and metadata module (e.g. x/IID) which could service all modules.
In working through this, Joe and I feel there are advantages to an asset module handling its own identifiers and metadata, within the context of the module and the services it offers. 
We would be keen to hear what others think about this. 
In the interim, we are continuing to develop the IID interface and a demo module. The next question would be whether we should submit another ADR to propose an IID module for the SDK.

Reply:
@aaronc I agree that renaming the "module" component to namespace is a solid improvement. There need not be the expectation that all namespaces within a chain are 1:1 controlled by modules.

That said, there is a nuance with IIDs that I want to anchor.
@aaronc wrote:
> It's not as if just any module exposes its own asset interface.

Correct, if a module isn't managing on-chain state, there's no need to expose an IID interface.

However, any module that manages on-chain state of any kind of asset, whether those assets are NFTs, FTs, accounts, services, transactions, encumbrances, or something else would benefit from IIDs. All a module needs to do is expose an IID interface (and provide the routing at the hub and chain level) and it can plug in to the cosmos IID ecosystem and have a fully conformant DID for whatever assets it manages. 

We anticipate, for example, that bespoke licensing modules are going to emerge to handle arbitrarily complex domain-specific encumbrances for NFTs that reside elsewhere (either on another module or another chain). Each of these modules would/could/should expose an IID for controlling their assets, e.g., the licensing contract attached to an NFT but managed by a module that understands that Paris is in France or the royalty agreement attached to a given NFT. These assets may or may not be "tokens" in any sense.

In other words, I see far more modules exposing an IID interface than I think you've been considering. It isn't just the NFTs and accounts that need IIDs; anything the manages an on-chain state and exposes an interface for managing that, can (and IMO should) expose an IID interface.

Of course, this transition will be gradual and it will likely be a while before it is the norm. However, the architecture of IIDs is far more flexible than your comments suggest.

@ig-shaun and I have been working through whether or not there is a place for an IID module that is, itself, managing the state of IID Documents by recording them, explicitly, on chain. My sensibility is that such a module is likely going to replicate state-management logic already encoded in the module that manages the asset. For example, if we had a cryptokitties module (which uses native cosmos modules for some of the underlying work, like bank and auth) that module has all the logic for translating on-chain records (based on signed transactions) to the "cattribute" properties that define a given cryptokitty, such as genes and "siredWith".

I imagine an IID module that provides helper functions to turn a set of properties and behaviors into well-formed IID Documents, but which itself is not managing state of those properties & behaviors. Alternatively, a cryptokitty module could defer (at least some) state management to an IID Module, but my estimate is that will lead to duplicate data stored on chain, which feels like both a systemic waste and an attack vector should the IID Module and the asset module get out of sync for any reason. We're still working through these tradeoffs.

Reply:
Thanks for sharing that context @jandrieu. My sense is that generally the IID layer shouldn't manage its own state, but rather act as a translation layer which exposes the IID API for existing modules.